### PR TITLE
Change schedulers to chainable form

### PR DESCRIPTION
**Pull request by vincentqb**
Enable chainable schedulers as requested in #13022 by implementing the changes mentioned below from [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513370208). The commit is self-contained and includes #21800.

Stack from [ghstack](https://github.com/ezyang/ghstack):

* Changing the behavior of schedulers to their chainable formula when supported
* Using the closed form whenever epoch is different from None until the next release with a deprecation warning
* Making get_computed_values the supported way of obtaining the last computed learning rate by the scheduler (see [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513940729) for new syntax)
* Returning a deprecation warning when invoking the undocumented get_lr function (see [comment](https://github.com/pytorch/pytorch/pull/21800#discussion_r294305485)) referring to get_computed_values, and deprecating it in the next release.
* `CosineAnnealingWarmRestart` still takes an epoch parameter as it is the only one with a mechanic relying on fractional epoch

### CODE REVIEW

**Comment by ezyang** [comment]
Can you rebase and resubmit this?

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -59,9 +78,26 @@ def load_state_dict(self, state_dict):
...
+
+    def _compute_values_closed_form(self):
+        # NOTE [LR Closed Form] Compute learning rate using closed form of the scheduler
         raise NotImplementedError
 
+    def get_lr(self):
+        warnings.warn(
+            "get_computed_values is now the supported method to obtain the last "
+            "computed learning rate. get_lr now returns the same, but will be "
+            "deprecated in favor of get_computed_values.", DeprecationWarning
```
Comment (ezyang) [inline torch/optim/lr_scheduler.py]: When I'm told that "instead of using `f`, I should use `g`" where they have exactly the same meaning, I tend to also like to know why a renaming happened at all. Is there actually a subtle semantics change? Is the old name bad and the new name more correctly says what happens? Just saying "foo is the new way of doing bar, bar returns the same thing" sounds like you've just been moving deck chairs around.

File: torch/optim/lr_scheduler.py [hunk 2]
```diff
@@ -1,23 +1,40 @@
...
+from functools import partial, wraps
 from bisect import bisect_right
+from torch._six import inf
 
 from .optimizer import Optimizer
 
 
+EPOCH_DEPRECATION_WARNING = ("The epoch parameter is being deprecated. If epoch "
+                             "is different from None, the closed form is used "
+                             "whenever available instead of the chainable form.")
```
Comment (ezyang) [inline torch/optim/lr_scheduler.py]: Yeah, I'm going to harp on deprecation warnings some more. Put yourself in the shoes of the user: they had some perfectly reasonable, working code on a previous version of PyTorch, and now it's throwing these warnings. What does the warning say? "Epoch is deprecated." OK. Why is it being deprecated? What should I use instead? "If epoch is different from None, the closed form is used whenever available instead of the chainable form." OK.... you didn't answer either of my questions!

File: torch/optim/lr_scheduler.py [hunk 3]
```diff
@@ -1,23 +1,40 @@
...
+        self._ensure_opt_first_helper()
+        self.step()
+
+    def _init_epoch_and_base(self, optimizer, last_epoch):
         if last_epoch == -1:
             for group in optimizer.param_groups:
                 group.setdefault('initial_lr', group['lr'])
-            last_epoch = 0
+            # last_epoch = 0
+            # step_last_epoch = None
```
Comment (ezyang) [inline torch/optim/lr_scheduler.py]: Kill these lines? (Unless you have a reason to keep them?)

File: torch/optim/lr_scheduler.py [hunk 4]
```diff
@@ -59,9 +78,26 @@ def load_state_dict(self, state_dict):
         """
         self.__dict__.update(state_dict)
 
-    def get_lr(self):
+    def get_computed_values(self):
+        # Return last computed learning rate by current scheduler
+        return self._last_computed_values
+
+    def _compute_values(self):
+        # NOTE [LR Chainable Form] Compute learning rate using chainable form of the scheduler
```
Comment (ezyang) [inline torch/optim/lr_scheduler.py]: So, the way the `Note [Foo]` convention works in the codebase, is that you have some long explanation that lives somewhere else, like:

```
Note [Foo]
~~~~~~~~~
blah blah blah
```

And then you reference the note from relevant code fragments `See Note [Foo]`. I don't see any such long explanation here. Do you want to add a longer explanation? Otherwise, just get rid of the `Note`

File: torch/optim/lr_scheduler.py [hunk 5]
```diff
@@ -26,6 +43,7 @@ def __init__(self, optimizer, last_epoch=-1):
         self.base_lrs = list(map(lambda group: group['initial_lr'], optimizer.param_groups))
         self.last_epoch = last_epoch
```
Comment (ezyang) [inline torch/optim/lr_scheduler.py]: Previously, it was not possible for `self.last_epoch` to be set to `-1`. With your changes, it is now set to `-1` when you pass in the default. This change to the invariants of members of LRScheduler was not documented in the PR. Can you explain more to me why this is sound? Because the `last_epoch` field is not underscored, it counts as a public member field, and so this is a very clear BC-breaking change.

File: torch/optim/lr_scheduler.py [hunk 6]
```diff
@@ -59,9 +78,26 @@ def load_state_dict(self, state_dict):
         """
         self.__dict__.update(state_dict)
 
-    def get_lr(self):
+    def get_computed_values(self):
+        # Return last computed learning rate by current scheduler
+        return self._last_computed_values
```
Comment (ezyang) [inline torch/optim/lr_scheduler.py]: So for example, a way `get_computed_values` differs from `get_lr` is that it's not valid to call `get_computed_values` on a LR scheduler which has not been stepped once. Maybe it would be safer and more obvious to our users if we raised an error in this case, instead of returning a list of NaNs (which may not be detected until much later when you're looking at your parameters and trying to figure out why they've all become NaNs...)

**Review by ezyang — CHANGES_REQUESTED** [review]
I want to accept this. I really do. But the patch as written just seems too risky given that the branch cut is happening in a few days. cc'ing @gchanan and @soumith for a second opinion.

File: torch/optim/lr_scheduler.py (hunk 3 above)
Comment (vincentqb) [reply torch/optim/lr_scheduler.py]: Thanks for pointing this out: I'll remove them.

File: torch/optim/lr_scheduler.py (hunk 4 above)
Comment (vincentqb) [reply torch/optim/lr_scheduler.py]: Thanks for the clarification. I'll keep the comment and remove the tag as you suggest.

File: torch/optim/lr_scheduler.py (hunk 6 above)
Comment (vincentqb) [reply torch/optim/lr_scheduler.py]: Yes, let's return an error instead if step has not been run.

File: torch/optim/lr_scheduler.py (hunk 1 above)
Comment (vincentqb) [reply torch/optim/lr_scheduler.py]: Good point. I'll add more details, like
```
      "The behavior of schedulers has been changed from their closed form to their "                                                                                                                                                     
      "chainable form. As such, the epoch parameter no longer applies, and is being "                                                                                                                                                    
      "deprecated. During the deprecation, if epoch is different from None, the closed "                                                                                                                                                 
      "form is used whenever available instead of the chainable form." 
```

File: torch/optim/lr_scheduler.py (hunk 2 above)
Comment (vincentqb) [reply torch/optim/lr_scheduler.py]: Thanks again for the feedback. I'll add the "why?" and the "then what?" like so
```
      "The behavior of schedulers has been changed from their closed form to their "                                                                                                                                                     
      "chainable form. As such, the epoch parameter no longer applies, and is being "                                                                                                                                                    
      "deprecated. The new approach is to invoke step without the epoch parameter "                                                                                                                                                      
      "during the iteration. During the deprecation, if epoch is different from None, "                                                                                                                                                  
      " the closed form is used whenever available instead of the chainable form."  
```

File: torch/optim/lr_scheduler.py (hunk 5 above)
Comment (vincentqb) [reply torch/optim/lr_scheduler.py]: The default value of `last_epoch` has not changed: it was and is `-1`. However, during the initialization, 

* now: `last_epoch` defaults to -1, `step` is called and increments `last_epoch` to 0. 
* before: `last_epoch` defaults to -1, changed to 0, and `step` is called and increments to 1.

All the tests though were calling `step` outside of initialization with the epoch parameter, thus overriding `last_epoch`. This indeed raises the question: is this change backward compatible? 

However, running test/test_optim.py with pytorch-nightly works, but leads to 5 test failures when removing `epoch` from `step` when you would expect the same answers as before.

```
diff --git a/test/test_optim.py b/test/test_optim.py
index e334962c9..2a365649a 100644
--- a/test/test_optim.py
+++ b/test/test_optim.py
@@ -565,7 +565,7 @@ class TestLRScheduler(TestCase):
 
         def old_pattern2():
             for e in range(epochs):
-                scheduler.step(e)
+                scheduler.step()
                 self.opt.step()
 
         self.assertWarnsRegex(old_pattern2, r'how-to-adjust-learning-rate')
@@ -599,7 +599,7 @@ class TestLRScheduler(TestCase):
 
         def old_pattern2():
             for e in range(epochs):
-                scheduler.step(e)
+                scheduler.step()
                 self.opt.step()
 
         self.assertWarnsRegex(old_pattern2, r'how-to-adjust-learning-rate')
@@ -627,7 +627,7 @@ class TestLRScheduler(TestCase):
 
         def old_pattern2():
             for e in range(epochs):
-                scheduler.step(e)
+                scheduler.step()
                 self.opt.step()
 
         self.assertWarnsRegex(old_pattern2, r'how-to-adjust-learning-rate')
@@ -1153,7 +1153,7 @@ class TestLRScheduler(TestCase):
         if isinstance(schedulers, _LRScheduler):
             schedulers = [schedulers]
         for epoch in range(epochs):
-            [scheduler.step(epoch) for scheduler in schedulers]
+            [scheduler.step() for scheduler in schedulers]
             for param_group, target in zip(self.opt.param_groups, targets):
                 self.assertAlmostEqual(target[epoch], param_group['lr'],
                                        msg='LR is wrong in epoch {}: expected {}, got {}'.format(
@@ -1184,7 +1184,7 @@ class TestLRScheduler(TestCase):
             targets.append([group['lr'] for group in self.opt.param_groups])
         self.setUp()
         for epoch in range(epochs):
-            scheduler.step(epoch)
+            scheduler.step()
             for i, param_group in enumerate(self.opt.param_groups):
                 self.assertAlmostEqual(targets[epoch][i], param_group['lr'],
                                        msg='LR is wrong in epoch {}: expected {}, got {}'.format(
@@ -1198,7 +1198,7 @@ class TestLRScheduler(TestCase):
                 if isinstance(scheduler, ReduceLROnPlateau):
                     scheduler.step(metrics[epoch])
                 else:
-                    scheduler.step(epoch)
+                    scheduler.step()
             if verbose:
                 print('epoch{}:\tlr={}'.format(epoch, self.opt.param_groups[0]['lr']))
             for param_group, target in zip(self.opt.param_groups, targets):
```

I would in fact argue that this fixes the initial stepping behavior. Thoughts?

File: test/test_optim.py [hunk 7]
```diff
@@ -1028,6 +1181,8 @@ def test_CosineAnnealingWarmRestarts_lr1(self):
                     T_i = int(T_mult) * T_i
                 targets[0] += [eta_min + (0.05 - eta_min) * (1 + math.cos(math.pi * T_cur / T_i)) / 2]
                 targets[1] += [eta_min + (0.5 - eta_min) * (1 + math.cos(math.pi * T_cur / T_i)) / 2]
+            # NOTE CosineAnnealingWarmRestarts requires step before
```
Comment (vincentqb) [inline test/test_optim.py]: See fix mentioned [here](https://github.com/pytorch/pytorch/pull/23833#pullrequestreview-271629279).

**Comment by vincentqb** [comment]
> Can you rebase and resubmit this?

Rebased in #24352
